### PR TITLE
Fix `function-url-quotes` false positives when URLs have modifiers

### LINKED ISSUE #8545: Fix `function-url-quotes` false positives when URLs have modifiers
### What minimal example or steps are needed to reproduce the bug?

```css
a {
	background: url( 'foo.svg' crossorigin(anonymous) );
}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "function-url-quotes": "never"
  }
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgIyTAGsBzAJwgFdMp4VKyAbAChQHIAzCCAOgGcAbiTYowFPnwhkAliWmZmSTBEwBPALZU+AShTaA3FgC+IADQgO0xjABySdXEQwAHvYAO1nmAlnwKyyQIIBiYKCjoIGSU1nwRdCFhYREc1GA40ioAtAyMmQCOlBA4MLEI4SCYMAIwZBFYYUbGvpCYAQBiUupIOEEAVpKYvrBufEEJ5Xw4qtaM8jhx5YzdJfNm9RNTM3OZLQGZk0pQSGRQCxFLxZN1mCZGQA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "latest",
  "stylelint-config-standard": "latest"
}
```

### What did you expect to happen?

No problems to be reported

### What actually happened?

The following was reported:

```
2:19-52 error Unexpected quotes around "url" function argument (function-url-quotes)
```

### Do you have a proposal to fix the bug?

The url token (`url( foo.svg )` without quotes) does not support url modifiers.
While the url function (`url( 'foo.svg' )` with quotes) does support these.

Stylelint should not report issues or try to fix url functions that contain url modifiers.

Relevant specs:
- https://www.w3.org/TR/css-values-4/#url-value
- https://drafts.csswg.org/css-values-5/#request-url-modifiers

### PR DESCRIPTION
**Pull request by immitsu**
> Which issue, if any, is this issue related to?

Closes #8545 

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/function-url-quotes/index.mjs [hunk 1]
```diff
@@ -158,24 +158,26 @@ const rule = (primary, secondaryOptions) => {
...
 			}
 
-			const hasQuotes = leftTrimmedArgs.startsWith("'") || leftTrimmedArgs.startsWith('"');
+			const hasQuotes = trimmedArgs.startsWith("'") || trimmedArgs.startsWith('"');
 
 			if (expectQuotes && hasQuotes) return;
 
 			if (!expectQuotes && !hasQuotes) return;
 
+			if (hasQuotes && trimmedArgs[0] !== trimmedArgs[trimmedArgs.length - 1]) return;
```
Comment (jeddy3) [inline lib/rules/function-url-quotes/index.mjs]: Let's extract `trimmedArgs[0] !== trimmedArgs[trimmedArgs.length - 1])` into a function called `hasModifier(args)`, making the code self-documenting for future maintainers.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@immitsu Thank you for working on this. It's looking good.

I've made a refactoring suggestion.

Let's also add a reject test for:

`never` 

```css
a { background: url(foo.svg crossorigin(anonymous) ); }
```

That is autofixed to:

```css
a { background: url("foo.svg" crossorigin(anonymous) ); }
```

File: lib/rules/function-url-quotes/index.mjs (hunk 1 above)
Comment (immitsu) [reply lib/rules/function-url-quotes/index.mjs]: I moved the check into `hasModifier(args)`, and in the next commit, I rewrote it. This was necessary for the fix

File: lib/rules/function-url-quotes/index.mjs [hunk 2]
```diff
@@ -155,35 +168,53 @@ const rule = (primary, secondaryOptions) => {
...
+			const modifierIndex = findModifierIndex(trimmedArgs, hasQuotes);
 
-			const messageType = expectQuotes ? 'expected' : 'rejected';
-			const reportIndex = index + args.length - leftTrimmedArgs.length;
-			const reportEndIndex = index + args.length;
+			if (!expectQuotes && !hasQuotes && !modifierIndex) return;
+
+			if (hasQuotes && modifierIndex) return;
+
+			const messageType = modifierIndex || expectQuotes ? 'expected' : 'rejected';
```
Comment (Mouvedia) [inline lib/rules/function-url-quotes/index.mjs]: question

Is this handling a case which would be invalid in the first place?
e.g. `url(foo.bar qux())`

Comment (immitsu) [reply lib/rules/function-url-quotes/index.mjs]: Yes, here is [the test case](https://github.com/stylelint/stylelint/pull/8702/files#diff-e0da576fc1f00dcfbcee4b417ec2c343b5cc89a77e4e832b958f258cdb403289R770) for the fix

Comment (Mouvedia) [reply lib/rules/function-url-quotes/index.mjs]: I would understand if the primary option was `always` but it's a reject test for `never`.
You are making assumptions for the fix, i.e. you could either remove the modifier or add the quotes.

Hence I think you should

1. just ignore it for `never` i.e. accept `code: 'a { background: url(foo.svg crossorigin(anonymous) ); }',`
2. add a reject case for `always` that fixes `code: 'a { background: url(foo.svg crossorigin(anonymous) ); }',` to `code: 'a { background: url("foo.svg" crossorigin(anonymous) ); }',`

That would be in line with I have done in stylelint-stylistic/stylelint-stylistic#26.

> I chose not to fix `@charset 'utf-8';` if `"string-quotes": "single"` because that would be unexpected and not on scope with the rule.

If there's a conflict with the primary option and the CSS is invalid just ignore it.

Comment (immitsu) [reply lib/rules/function-url-quotes/index.mjs]: Updated.
Thank you for the explanation!

File: lib/rules/function-url-quotes/index.mjs [hunk 3]
```diff
@@ -155,35 +168,53 @@ const rule = (primary, secondaryOptions) => {
...
+				modifierIndex,
 			};
 		}
+
+		/**
+		 * @param {string} args
+		 * @param {boolean} startsWithQuote
+		 * @returns {number | undefined}
+		 */
+		function findModifierIndex(args, startsWithQuote) {
```
Comment (Mouvedia) [inline lib/rules/function-url-quotes/index.mjs]: note

We currently don't have test cases that use `var()` hence I suppose we don't support the indirection. You don't have to change anything; we should document that in the README though, in a subsequent documentation PR that is.
i.e. `v` !== `'` nor `"`

**Review by jeddy3 — APPROVED** [review]
Upon revisiting this PR, it seems I sent us off in the wrong direction with my request for autofix in https://github.com/stylelint/stylelint/pull/8702#pullrequestreview-3082963723. Sorry about that!

As per https://github.com/stylelint/stylelint/issues/8545#issue-3015138754, the rule:

> should not report issues or try to fix url functions that contain url modifiers.

As quotes aren't optional when a modifier is used, and unquoted URLs with a modifier are caught by [`declaration-property-value-no-unknown`](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown).

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYKBGSwGsDmAThAK4B2U8KxBANgBQBmEEAdAM4BueKYRbbEAgEs8Q0nSSkIpAJ4BbEmwCUSgNwoAviAA0IBkJowAckjlxEMAB6mADoZZh+O8NP14EIKzcEAXFLAYkYho-YAAdUhQUKx8YcjZKAG0AcjYfGUMaMR8AWkhSNxy0ySgkAihkgF1tCI1VCOd8twAxQTkkHw80jJgs0h8HVxEWOQArNmdYGwnEcMiUMJBuzOzFykWaDpg0xZr5xeXe7LyhvCKfErKoNYWQTdidkFqGjSA).

I've pushed a couple of commits to revert the autofix-related changes, and one to refactor the fix to check the AST for a URL modifier.

File: lib/rules/function-url-quotes/index.mjs [hunk 4]
```diff
@@ -221,6 +223,18 @@ const rule = (primary, secondaryOptions) => {
...
+
+		/**
+		 * @param {FunctionNode} funcNode
+		 * @returns {boolean}
+		 */
+		function hasUrlModifier(funcNode) {
+			return funcNode.nodes.some(
+				(n) =>
+					(n.type !== 'string' && n.type !== 'space' && n.type !== 'word') ||
+					(n.type === 'word' && /\s/.test(n.value)),
```
Comment (romainmenke) [inline lib/rules/function-url-quotes/index.mjs]:
```suggestion
					(n.type === 'word' && /(?:^|[^\\])\s/.test(n.value)),
```

Escaped whitespace are allowed in url tokens, so we should adjust the regex and add a test to cover that:

```js
		{
			code: '@import url(foo\\ .css);',
			fixed: '@import url("foo\\ .css");',
			fix: {
				range: [12, 21],
				text: '"foo\\ .css"',
			},
			message: messages.expected('url'),
			line: 1,
			column: 13,
			endLine: 1,
			endColumn: 22,
		},
```

**Comment by jeddy3** [comment]
@romainmenke Thank you for catching that edge case! I've committed your suggestion and pushed a commit for the corresponding test case.
